File: tests/cpp/test_musicxml.cpp
// Gingo — Music Theory Library
// Tests for MusicXML serializer.
//
// SPDX-License-Identifier: MIT

#include <catch2/catch_test_macros.hpp>
#include <gingo/musicxml.hpp>

#include <filesystem>

using namespace gingo;

// ---------------------------------------------------------------------------
// Note
// ---------------------------------------------------------------------------

TEST_CASE("MusicXML note C4", "[musicxml]") {
    auto xml = MusicXML::note(Note("C"), 4);
    REQUIRE(xml.find("<step>C</step>") != std::string::npos);
    REQUIRE(xml.find("<octave>4</octave>") != std::string::npos);
    REQUIRE(xml.find("<type>quarter</type>") != std::string::npos);
    REQUIRE(xml.find("<score-partwise") != std::string::npos);
    REQUIRE(xml.find("</score-partwise>") != std::string::npos);
}

TEST_CASE("MusicXML note with sharp", "[musicxml]") {
    auto xml = MusicXML::note(Note("C#"), 4);
    REQUIRE(xml.find("<step>C</step>") != std::string::npos);
    REQUIRE(xml.find("<alter>1</alter>") != std::string::npos);
}

TEST_CASE("MusicXML note with flat", "[musicxml]") {
    auto xml = MusicXML::note(Note("Bb"), 4);
    REQUIRE(xml.find("<step>B</step>") != std::string::npos);
    REQUIRE(xml.find("<alter>-1</alter>") != std::string::npos);
}

TEST_CASE("MusicXML note natural has no alter", "[musicxml]") {
    auto xml = MusicXML::note(Note("G"), 5);
    REQUIRE(xml.find("<alter>") == std::string::npos);
}

TEST_CASE("MusicXML note whole type", "[musicxml]") {
    auto xml = MusicXML::note(Note("E"), 3, "whole");
    REQUIRE(xml.find("<type>whole</type>") != std::string::npos);
}

// ---------------------------------------------------------------------------
// Chord
// ---------------------------------------------------------------------------

TEST_CASE("MusicXML chord CM", "[musicxml]") {
    auto xml = MusicXML::chord(Chord("CM"), 4);
    // First note is NOT a chord member, subsequent are.
    // Count <note> elements.
    std::size_t count = 0;
    std::size_t pos = 0;
    while ((pos = xml.find("<note>", pos)) != std::string::npos) {
        count++;
        pos++;
    }
    REQUIRE(count == 3);  // C, E, G

    // Second and third notes should have <chord/>.
    std::size_t chord_count = 0;
    pos = 0;
    while ((pos = xml.find("<chord/>", pos)) != std::string::npos) {
        chord_count++;
        pos++;
    }
    REQUIRE(chord_count == 2);
}

TEST_CASE("MusicXML chord Am7 has 4 notes", "[musicxml]") {
    auto xml = MusicXML::chord(Chord("Am7"), 4);
    std::size_t count = 0;
    std::size_t pos = 0;
    while ((pos = xml.find("<note>", pos)) != std::string::npos) {
        count++;
        pos++;
    }
    REQUIRE(count == 4);
}

// ---------------------------------------------------------------------------
// Scale
// ---------------------------------------------------------------------------

TEST_CASE("MusicXML scale C major", "[musicxml]") {
    auto xml = MusicXML::scale(Scale("C", ScaleType::Major), 4);
    // 7 notes in the scale.
    std::size_t count = 0;
    std::size_t pos = 0;
    while ((pos = xml.find("<note>", pos)) != std::string::npos) {
        count++;
        pos++;
    }
    REQUIRE(count == 7);

    // No <chord/> tags (sequential notes).
    REQUIRE(xml.find("<chord/>") == std::string::npos);
}

// ---------------------------------------------------------------------------
// Field
// ---------------------------------------------------------------------------

TEST_CASE("MusicXML field C major has 7 measures", "[musicxml]") {
    auto xml = MusicXML::field(Field("C", ScaleType::Major), 4);
    std::size_t count = 0;
    std::size_t pos = 0;
    while ((pos = xml.find("<measure number=", pos)) != std::string::npos) {
        count++;
        pos++;
    }
    REQUIRE(count == 7);
}

// ---------------------------------------------------------------------------
// Sequence
// ---------------------------------------------------------------------------

TEST_CASE("MusicXML sequence basic", "[musicxml]") {
    Sequence seq(Tempo(120), TimeSignature(4, 4));
    seq.add(NoteEvent(Note("C"), Duration("quarter"), 4));
    seq.add(NoteEvent(Note("E"), Duration("quarter"), 4));
    seq.add(Rest(Duration("half")));

    auto xml = MusicXML::sequence(seq);
    REQUIRE(xml.find("<score-partwise") != std::string::npos);
    REQUIRE(xml.find("<step>C</step>") != std::string::npos);
    REQUIRE(xml.find("<step>E</step>") != std::string::npos);
    REQUIRE(xml.find("<rest/>") != std::string::npos);
}

// ---------------------------------------------------------------------------
// Write
// ---------------------------------------------------------------------------

TEST_CASE("MusicXML write to file", "[musicxml]") {
    auto xml = MusicXML::note(Note("C"), 4);
    std::string path = "/tmp/gingo_test_musicxml.xml";
    MusicXML::write(xml, path);
    REQUIRE(std::filesystem::exists(path));
    std::filesystem::remove(path);
}

// ---------------------------------------------------------------------------
// Header/footer structure
// ---------------------------------------------------------------------------

TEST_CASE("MusicXML valid document structure", "[musicxml]") {
    auto xml = MusicXML::note(Note("D"), 4);
    REQUIRE(xml.find("<?xml version=") != std::string::npos);
    REQUIRE(xml.find("<!DOCTYPE score-partwise") != std::string::npos);
    REQUIRE(xml.find("<software>Gingo</software>") != std::string::npos);
    REQUIRE(xml.find("<part-name>Piano</part-name>") != std::string::npos);
    REQUIRE(xml.find("<divisions>") != std::string::npos);
}

File: tests/python/test_musicxml.py
"""Tests for MusicXML serialization."""

import os
import pytest
from gingo import (
    MusicXML, Note, Chord, Scale, ScaleType, Field,
    Sequence, Tempo, TimeSignature, NoteEvent, ChordEvent, Rest, Duration,
)


# ---------------------------------------------------------------------------
# Note
# ---------------------------------------------------------------------------

class TestMusicXMLNote:
    def test_basic_note(self):
        xml = MusicXML.note(Note("C"), 4)
        assert "<step>C</step>" in xml
        assert "<octave>4</octave>" in xml
        assert "<type>quarter</type>" in xml

    def test_sharp(self):
        xml = MusicXML.note(Note("C#"), 4)
        assert "<step>C</step>" in xml
        assert "<alter>1</alter>" in xml

    def test_flat(self):
        xml = MusicXML.note(Note("Bb"), 4)
        assert "<step>B</step>" in xml
        assert "<alter>-1</alter>" in xml

    def test_natural_no_alter(self):
        xml = MusicXML.note(Note("G"), 5)
        assert "<alter>" not in xml

    def test_whole_type(self):
        xml = MusicXML.note(Note("E"), 3, "whole")
        assert "<type>whole</type>" in xml

    def test_document_structure(self):
        xml = MusicXML.note(Note("D"), 4)
        assert '<?xml version=' in xml
        assert '<!DOCTYPE score-partwise' in xml
        assert '<score-partwise' in xml
        assert '</score-partwise>' in xml
        assert '<software>Gingo</software>' in xml
        assert '<part-name>Piano</part-name>' in xml


# ---------------------------------------------------------------------------
# Chord
# ---------------------------------------------------------------------------

class TestMusicXMLChord:
    def test_cm_three_notes(self):
        xml = MusicXML.chord(Chord("CM"), 4)
        assert xml.count("<note>") == 3

    def test_cm_chord_tags(self):
        xml = MusicXML.chord(Chord("CM"), 4)
        # 2nd and 3rd notes get <chord/>
        assert xml.count("<chord/>") == 2

    def test_am7_four_notes(self):
        xml = MusicXML.chord(Chord("Am7"), 4)
        assert xml.count("<note>") == 4


# ---------------------------------------------------------------------------
# Scale
# ---------------------------------------------------------------------------

class TestMusicXMLScale:
    def test_c_major_seven_notes(self):
        xml = MusicXML.scale(Scale("C", ScaleType.Major), 4)
        assert xml.count("<note>") == 7

    def test_no_chord_tags(self):
        xml = MusicXML.scale(Scale("C", ScaleType.Major), 4)
        assert "<chord/>" not in xml


# ---------------------------------------------------------------------------
# Field
# ---------------------------------------------------------------------------

class TestMusicXMLField:
    def test_seven_measures(self):
        xml = MusicXML.field(Field("C", ScaleType.Major), 4)
        assert xml.count("<measure number=") == 7


# ---------------------------------------------------------------------------
# Sequence
# ---------------------------------------------------------------------------

class TestMusicXMLSequence:
    def test_basic_sequence(self):
        seq = Sequence(Tempo(120), TimeSignature(4, 4))
        seq.add(NoteEvent(Note("C"), Duration("quarter"), 4))
        seq.add(NoteEvent(Note("E"), Duration("quarter"), 4))
        seq.add(Rest(Duration("half")))

        xml = MusicXML.sequence(seq)
        assert "<step>C</step>" in xml
        assert "<step>E</step>" in xml
        assert "<rest/>" in xml

    def test_chord_event(self):
        seq = Sequence(Tempo(120), TimeSignature(4, 4))
        seq.add(ChordEvent(Chord("CM"), Duration("whole"), 4))

        xml = MusicXML.sequence(seq)
        assert xml.count("<note>") == 3
        assert xml.count("<chord/>") == 2

    def test_tempo_marking(self):
        seq = Sequence(Tempo(120), TimeSignature(4, 4))
        seq.add(NoteEvent(Note("C"), Duration("quarter"), 4))

        xml = MusicXML.sequence(seq)
        assert "<per-minute>120</per-minute>" in xml


# ---------------------------------------------------------------------------
# Write
# ---------------------------------------------------------------------------

class TestMusicXMLWrite:
    def test_write_file(self, tmp_path):
        xml = MusicXML.note(Note("C"), 4)
        path = str(tmp_path / "test.musicxml")
        MusicXML.write(xml, path)
        assert os.path.exists(path)
        with open(path) as f:
            content = f.read()
        assert "<step>C</step>" in content

    def test_write_chord(self, tmp_path):
        xml = MusicXML.chord(Chord("Am7"), 4)
        path = str(tmp_path / "chord.musicxml")
        MusicXML.write(xml, path)
        assert os.path.exists(path)
